### PR TITLE
Column compression cleanup

### PR DESCRIPTION
**Pull request by rxin**
- In dictionary encoding, the dictionary itself doesn't need to store the index (the values are stored in order of the index)
- In dictionary encoding decoder, uses an array to replace the original hash table so dictionary lookup is always O(1)
- Remove the position argument for extract/extractInfo methods in ColumnType
- init() is called in a ColumnIterator's constructor, and removed computeNext()
- Corrected various tests. Some tests were testing for the wrong thing (e.g. a few tests used null count for compression type ...)
- Various other minor changes to improve code readability and test readability

### CODE REVIEW

File: src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala [hunk 1]
```diff
@@ -12,117 +12,116 @@ import shark.memstore2.column.Implicits._
...
+      t.append(item, b)
+      algo.gatherStatsForCompressibility(item, t.asInstanceOf[ColumnType[Any, _]])
     }
     b.limit(b.position())
     b.rewind()
-    val compressedBuffer = algo.compress(b, u)
+
     if (shouldNotCompress) {
       assert(algo.compressionRatio >= 1.0)
-      info("CompressionRatio " + algo.compressionRatio)
```
Comment (junkumar) [inline src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala:44]: Why remove this information from the test output?

File: src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala [hunk 2]
```diff
@@ -12,117 +12,116 @@ import shark.memstore2.column.Implicits._
...
+        case _       => throw new UnsupportedOperationException("Unsupported compression type " + t)
       }
 
       l.foreach { x =>
         iter.next()
-        assert(compareFunc(u.get(iter.current, oi), x))
-        // assert(u.get(iter.current, oi) === x)
+        assert(compareFunc(t.get(iter.current, oi), x))
       }
-      assert(false === iter.hasNext) // no extras at the end
```
Comment (junkumar) [inline src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala:63]: I think the triple = gives a better message upon failure. Just checking the function's return dumps a harder to read stack trace on the console.

File: src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala [hunk 3]
```diff
@@ -12,117 +12,116 @@ import shark.memstore2.column.Implicits._
...
+      shouldNotCompress = true)
   }
 
   test("RLE - should not compress") {
-    val alternating = Range(0, Short.MaxValue-1, 1).flatMap { s => List(1, s) }
-    val hugeList = List.concat(iList, alternating, iList)
-    assert(hugeList.size === (8 + 2*(Short.MaxValue-1)))
-    testList(hugeList, INT, new RLE(), (a: Int, b: Int) => a == b, true)
+    val ints = Range(0, Short.MaxValue.toInt + 1)
+    testList(ints, INT, new RLE, (a: Int, b: Int) => a == b, shouldNotCompress = true)
```
Comment (junkumar) [inline src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala]: How about keeping the (shouldNotCompress = true) test at (Short.MaxValue.toInt + 1) as well as the 
 (shouldNotCompress = false) test at (Short.MaxValue.toInt - 1)

This way we test the limit from both directions.

File: src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala (hunk 2 above)
Comment (rxin) [inline src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala:63]: It tells you a message like "true does not equal false". I don't think that is more helpful than the assertion is not true ...

File: src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala (hunk 3 above)
Comment (rxin) [inline src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala]: Good idea. (Actually I still have the (Short.MaxValue.toInt - 1) test. Look about 15 lines up)

File: src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala (hunk 1 above)
Comment (rxin) [inline src/test/scala/shark/memstore2/column/CompressedColumnIteratorSuite.scala:44]: Actually good idea - I am adding this back

**Pull request merged**